core/src/main/java/org/apache/iceberg/util/Tasks.java

```diff
@@ -213,7 +214,7 @@ public boolean run(Task<I, RuntimeException> task) {
           try {
             runTaskWithRetry(task, item);
             succeeded.add(item);
-
+            threw = false;
```

Comment:
I believe we need to set `threw = false` here instead of in L230, as otherwise `threw` will always be `false` when it reaches the check in https://github.com/apache/iceberg/pull/3215/files#diff-e075c62463ddc17c45f8821ee302c98b4ef0c74c304d27f5e82d08470594bce6R271

Reply:
I'm not sure I follow, 

We have

While (items) 
  try to do something
  catch failure and try to handle
done
No breakthrough exceptions for any items threw = false

Reply:
that makes sense, but then I believe we can simplify the code to the one below as `throwFailureWhenFinished && threw` is always `false` according to the IDE?

```
--- a/core/src/main/java/org/apache/iceberg/util/Tasks.java
+++ b/core/src/main/java/org/apache/iceberg/util/Tasks.java
@@ -270,12 +270,9 @@ public class Tasks {
 
       if (throwFailureWhenFinished && !exceptions.isEmpty()) {
         Tasks.throwOne(exceptions, exceptionClass);
-      } else if (throwFailureWhenFinished && threw) {
-        throw new RuntimeException(
-            "Task set failed with an uncaught throwable");
       }
 
-      return !threw;
+      return true;

```

Reply:
> This block of code is quite confusing :)

I agree, that's what made me look at it :)

I'll leave this block of code for now as it is and will investigate a bit more